consensus/src/dag/tests/dag_state_sync_tests.rs

```diff
@@ -0,0 +1,216 @@
+// Copyright © Aptos Foundation
+
+use crate::{
+    dag::{
+        adapter::Notifier,
+        dag_fetcher::{FetchRequestHandler, TDagFetcher},
+        dag_state_sync::{StateSyncManager, DAG_WINDOW},
+        dag_store::Dag,
+        storage::DAGStorage,
+        tests::{dag_test::MockStorage, helpers::generate_dag_nodes},
+        types::{CertifiedNodeMessage, RemoteFetchRequest},
+        CertifiedNode, DAGMessage, RpcHandler, RpcWithFallback, TDAGNetworkSender,
+    },
+    test_utils::EmptyStateComputer,
+};
+use aptos_consensus_types::common::{Author, Round};
+use aptos_crypto::HashValue;
+use aptos_infallible::RwLock;
+use aptos_reliable_broadcast::RBNetworkSender;
+use aptos_time_service::TimeService;
+use aptos_types::{
+    aggregate_signature::AggregateSignature,
+    block_info::BlockInfo,
+    epoch_change::EpochChangeProof,
+    epoch_state::EpochState,
+    ledger_info::{LedgerInfo, LedgerInfoWithSignatures},
+    validator_verifier::random_validator_verifier,
+};
+use async_trait::async_trait;
+use claims::assert_none;
+use std::{sync::Arc, time::Duration};
+
+struct MockDAGNetworkSender {}
+
+#[async_trait]
+impl RBNetworkSender<DAGMessage> for MockDAGNetworkSender {
+    async fn send_rb_rpc(
+        &self,
+        _receiver: Author,
+        _message: DAGMessage,
+        _timeout: Duration,
+    ) -> anyhow::Result<DAGMessage> {
+        unimplemented!()
+    }
+}
+
+#[async_trait]
+impl TDAGNetworkSender for MockDAGNetworkSender {
+    async fn send_rpc(
+        &self,
+        _receiver: Author,
+        _message: DAGMessage,
+        _timeout: Duration,
+    ) -> anyhow::Result<DAGMessage> {
+        unimplemented!()
+    }
+
+    /// Given a list of potential responders, sending rpc to get response from any of them and could
+    /// fallback to more in case of failures.
+    async fn send_rpc_with_fallbacks(
+        &self,
+        _responders: Vec<Author>,
+        _message: DAGMessage,
+        _retry_interval: Duration,
+        _rpc_timeout: Duration,
+    ) -> RpcWithFallback {
+        unimplemented!()
+    }
+}
+
+struct MockDagFetcher {
+    target_dag: Arc<RwLock<Dag>>,
+    epoch_state: Arc<EpochState>,
+}
+
+#[async_trait]
+impl TDagFetcher for MockDagFetcher {
+    async fn fetch(
+        &self,
+        remote_request: RemoteFetchRequest,
+        _responders: Vec<Author>,
+        new_dag: Arc<RwLock<Dag>>,
```

Comment:
nit: I think when we do fetch it can be a Dag without arc mutex?

Reply:
yes, but the same trait is used for fetcher service too, which needs to pass an arc mutex.